compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/impl/FirPartiallyResolvedArgumentList.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.expressions.impl
+
+import org.jetbrains.kotlin.fir.FirSourceElement
+import org.jetbrains.kotlin.fir.declarations.FirValueParameter
+import org.jetbrains.kotlin.fir.expressions.FirArgumentList
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.fir.visitors.FirTransformer
+import org.jetbrains.kotlin.fir.visitors.FirVisitor
+import org.jetbrains.kotlin.fir.visitors.transformInplace
+
+class FirPartiallyResolvedArgumentList internal constructor(
+    override var source: FirSourceElement?,
+    override val arguments: MutableList<FirExpression>,
+    private val matchedValueParameters: List<FirValueParameter?>
+) : FirArgumentList() {
+
+    val mapping: Map<FirExpression, FirValueParameter>
+        get() {
+            return matchedValueParameters.mapIndexedNotNull { index, firValueParameter ->
+                arguments[index] to (firValueParameter ?: return@mapIndexedNotNull null)
+            }.toMap()
+        }
+
+    override fun <R, D> acceptChildren(visitor: FirVisitor<R, D>, data: D) {
+        arguments.forEach { it.accept(visitor, data) }
+    }
+
+    override fun <D> transformChildren(transformer: FirTransformer<D>, data: D): FirPartiallyResolvedArgumentList {
+        transformArguments(transformer, data)
+        return this
+    }
+
+    override fun <D> transformArguments(transformer: FirTransformer<D>, data: D): FirPartiallyResolvedArgumentList {
+        arguments.transformInplace(transformer, data)
```

Comment:
See https://github.com/JetBrains/kotlin/pull/3924#discussion_r529265316

We may need something similar to `FirResolvedArgumentList#transformArguments`

Reply:
I think this should be fine since the transformation on arguments indeed happens in place. I intentionally keyed the mapping with index rather than `FirExpression` so that the mapping can stay unchanged during transformation

Reply:
Actually, following Marks' comment, I am changing the implementation to be backed by a `LinkedHashMap` instead.